### DISCUSSION TITLE
Not working Qt6 Qt::ColorScheme

### DISCUSSION THREAD

**Discussion by redtide** [tsujan/Kvantum#921] [Q&A] [answered]
Hello,
I'm trying to update a widgets set project with the possibility to react on theme changes.

I see that from 6.5, Qt has a new [colorScheme](https://doc.qt.io/qt-6/qstylehints.html#colorScheme-prop) property, to know which color scheme is currently in use; `Dark`, `Light` or `Unknown` ([here](https://www.qt.io/blog/dark-mode-on-windows-11-with-qt-6.5) an interesting article in Windows 11 context).

Unfortunately in LXQt it seems to return always `Unknown`, so I wanted to ask you, as Kvantum and LXQt developer, if there is something that one of them can do about it, or if it's a Qt (XCB at least) non working feature for some reason and/or context.

For now it seems I get successful results by using a function that checks the relative luminance of `QPalette::Window`' color, together with the (undocumented) `QEvent::ThemeChange` event type, which works also with Qt5.

**Comment by tsujan** [comment, marked as answer]
It's because the Qt6 branch of `lxqt-qtplugin` (provided that you have it) doesn't support `colorScheme()` yet and Qt → `qplatformtheme.cpp` contains

```c++
Qt::ColorScheme QPlatformTheme::colorScheme() const
{
    return Qt::ColorScheme::Unknown;
}
```

That being said, even when we add it to the Qt6 branch of `lxqt-qtplugin`, it won't be applicable with style plugins like Kvantum, which have their own palette and don't follow the platform palette.

**Reply by redtide** [reply]
Thank you for the response, though I don't understand why it shouldn't work with Kvantum since it's about only to return if the scheme is dark or not, information that it has, despite the different palette.
I guess I can continue to use my own method if not to find something better.

**Reply by tsujan** [reply]
> though I don't understand why it shouldn't work with Kvantum

It'll return the scheme of the platform's palette when implemented, i.e., what's in LXQt Appearance Configuration → Qt Palette (or its counterpart in KDE). For example, the platform's palette may be light, but the active Kvantum theme may be dark, and conversely.

Kvantum makes sure that its own palette is applied to the app, instead of the platform palette, becauset each Kvantum theme has SVG elements that match only the specific colors defined in its `kvconfig` file (otherwise, many properties of Kvantum, which are beyond what Qt provides, wouldn't be possible — like a dark toolbar or a dark menu background with a light theme).

**Reply by tsujan** [reply]
To be more precise, I should say that it'll really return the scheme of what's in `~/.config/lxqt/lxqt.conf` (or, in the absence of any palette in that file, the scheme of Qt's default theme, Fusion).

If there's no palette there, LXQt Appearance Configuration will show the palette of Kvantum's active theme (which is the palette of the dialog of LXQt Appearance Configuration); otherwise, it'll show what's in that file.

Frankly, the situation can be more complicated than this. The good thing about Qt's styling is that it's very flexible; the bad thing is that Qt devs may forget how flexible it is and write inconsistent codes — to say nothing of devs who write Qt-based apps and include wrong style codes in them.

**Reply by redtide** [reply]
FWIW I could be wrong but all this sounds too complicated and odd to me.
What I did and that makes sense to me it's to choose KvGnome in Kvantum as base theme, which has also a dark counterpart, then choose kvantum-dark in lxqt-config-appearance to select KvGnomeDark as consequence, which is a dark theme, so I would expect Qt to return a Dark color scheme. IDK about the custom palettes, but it would make sense that for people like me that have no interest to customize theme' palettes, to use the one offered by the theme so the system knows what's going on and return the appropriate scheme without caring to customized settings, and, if any then be invalidated, not return nothing because it's all a customized hell.

**Reply by tsujan** [reply]
> ... so I would expect Qt to return a Dark color scheme

As far as I've checked the code of Qt 6.6, I think that shouldn't happen, because Qt consults the platform, not the actual palette. And yes, it's an inconsistency (but not the only one I've encountered in Qt).

My guess is that many Qt devs don't know of possibilities beyond Fusion-like themes.

**Reply by redtide** [reply]
I have no much an idea how it works in Qt, ideally to me Kvantum should tell to LXQt plugin which color mode is using (from theme or custom palette) and then the plugin report it to Qt, but IDK how this should be done.

From code I see in `src/plugins/platformthemes` there are a gtk3 and a xdgdesktopportal plugins, where IIUC gtk3 search for a ["dark" value substring](https://github.com/qt/qtbase/blob/dev/src/plugins/platformthemes/gtk3/qgtk3storage.cpp#L340) in the [gtk-theme-name](https://github.com/qt/qtbase/blob/dev/src/plugins/platformthemes/gtk3/qgtk3interface.cpp#L484) key on `$HOME/.config/gtk3.0/settings.ini` file,  and the `color-scheme` [key](https://github.com/qt/qtbase/blob/dev/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp#L120) in [org.freedesktop.portal.Settings](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Settings.html).

Something similar to GTK3 is made also for [KDE](https://github.com/qt/qtbase/blob/ca38ac77313cd4ad1436fae28ec43dc219ea341d/src/gui/platform/unix/qgenericunixthemes.cpp#L961).

**Reply by tsujan** [reply]
Being a style plugin, Kvantum reports what `QCommonStyle` does. It's up to Qt devs to add a method to `QStyle` for getting a "Qt::ColorScheme".

Breeze and Fusion are irrelevant as far as Kvantum is concerned, let alone GTK3.

**Reply by redtide** [reply]
Thank you for your time and clarifications. Should I close this?

**Reply by tsujan** [reply]
You're very welcome.

As this contains information that may be useful to others (well, it was useful to *me*), I'd prefer to keep it open. Generally, discussions don't need to be closed.
